cpp/src/arrow/dataset/discovery.cc

```diff
@@ -58,57 +62,88 @@ bool StartsWithAnyOf(const std::vector<std::string>& prefixes, const std::string
   return std::any_of(prefixes.cbegin(), prefixes.cend(), matches_prefix);
 }
 
-Result<DataSourceDiscoveryPtr> FileSystemDataSourceDiscovery::Make(
-    fs::FileSystemPtr fs, fs::FileStatsVector files, FileFormatPtr format,
-    FileSystemDiscoveryOptions options) {
-  DCHECK_NE(format, nullptr);
-
-  bool has_prefixes = !options.ignore_prefixes.empty();
-  std::vector<fs::FileStats> filtered;
-  for (const auto& stat : files) {
-    if (stat.IsFile()) {
-      const std::string& path = stat.path();
+Status FileSystemDataSourceDiscovery::Filter(const fs::FileSystemPtr& filesystem,
+                                             const FileFormatPtr& format,
+                                             const FileSystemDiscoveryOptions& options,
+                                             fs::FileStatsVector* files) {
+  for (auto it = files->begin(); it != files->end();) {
+    if (StartsWithAnyOf(options.ignore_prefixes, it->path())) {
+      files->erase(it);
+      continue;
+    }
 
-      if (has_prefixes && StartsWithAnyOf(options.ignore_prefixes, path)) {
+    if (it->IsFile() && options.exclude_invalid_files) {
+      ARROW_ASSIGN_OR_RAISE(
+          auto supported, format->IsSupported(FileSource(it->path(), filesystem.get())));
+      if (!supported) {
+        files->erase(it);
         continue;
       }
+    }
+
+    ++it;
+  }
+  return Status::OK();
+}
+
+Result<DataSourceDiscoveryPtr> FileSystemDataSourceDiscovery::Make(
+    fs::FileSystemPtr filesystem, const std::vector<std::string>& paths,
+    FileFormatPtr format, FileSystemDiscoveryOptions options) {
+  DCHECK_NE(format, nullptr);
+
+  ARROW_ASSIGN_OR_RAISE(auto files, filesystem->GetTargetStats(paths));
+  ARROW_ASSIGN_OR_RAISE(auto forest, fs::PathForest::Make(std::move(files)));
 
-      if (options.exclude_invalid_files) {
-        ARROW_ASSIGN_OR_RAISE(auto supported,
-                              format->IsSupported(FileSource(path, fs.get())));
-        if (!supported) {
-          continue;
-        }
+  std::vector<std::string> missing;
```

Comment:
Use `unordered_set` instead?

Reply:
I don't see an advantage over vector; uniqueness is guaranteed, they're both O(1) insert, and vector is slightly friendlier to the allocator

Reply:
Ah, I see: missing segments are guaranteed within a single path but not between paths. I'll use the set